avro-fastserde/src/main/java/com/linkedin/avro/fastserde/primitive/PrimitiveArrayList.java

```diff
@@ -0,0 +1,137 @@
+package com.linkedin.avro.fastserde.primitive;
+
+import java.util.AbstractList;
+import java.util.Iterator;
+import org.apache.avro.generic.GenericArray;
+import org.apache.avro.generic.GenericContainer;
+import org.apache.avro.generic.GenericData;
+
+
+public abstract class PrimitiveArrayList<T, L> extends AbstractList<T>
+    implements GenericContainer, Comparable<GenericArray<T>> {
+  private final Class<L> type;
+  protected int size = 0;
+
+  public PrimitiveArrayList(Class<L> type) {
+    this.type = type;
+  }
+
+  @Override public int size() {
+    return size;
+  }
+
+  @Override public void clear() {
+    size = 0;
+  }
+
+  @Override public Iterator<T> iterator() {
+    return new Iterator<T>() {
+      private int position = 0;
+
+      @Override
+      public boolean hasNext() {
+        return position < size;
+      }
+
+      @Override
+      public T next() {
+        return get(position++);
+      }
+
+      @Override
+      public void remove() {
+        throw new UnsupportedOperationException();
+      }
+    };
+  }
+
+  protected abstract int capacity();
+
+  protected void checkIfLargerThanSize(int i) {
+    if (i >= size)
+      throw new IndexOutOfBoundsException("Index " + i + " out of bounds.");
+  }
+
+  protected void capacityCheck() {
+    if (size == capacity()) {
+      Object newElements = newArray((size * 3)/2 + 1);
+      System.arraycopy(getElementsArray(), 0, newElements, 0, size);
+      setElementsArray(newElements);
+    }
+  }
+
+  protected abstract void setElementsArray(Object newElements);
+
+  protected abstract Object getElementsArray();
+
+  protected abstract Object newArray(int capacity);
+
+  /** Create an empty space in the array at the index specified by {@param location} by shifting the elements to the right */
+  protected void addInternal(int location) {
```

Comment:
Please carefully review this function. I've tweaked the array resize/copy code, compared to `GenericData.Array`... I think I got it right but please apply a good dose of scrutiny.

Reply:
lgtm